paimon-core/src/main/java/org/apache/paimon/table/RollbackHelper.java

```diff
@@ -0,0 +1,191 @@
+    private final SnapshotDeletion snapshotDeletion;
+    private final TagDeletion tagDeletion;
+
+    private final SortedMap<Snapshot, String> tags;
+    private final Map<Snapshot, List<ManifestEntry>> tagManifestEntries;
+
+    private Set<String> manifestsSkippingSet;
+
+    public RollbackHelper(
+            SnapshotManager snapshotManager,
+            TagManager tagManager,
+            FileIO fileIO,
+            SnapshotDeletion snapshotDeletion,
+            TagDeletion tagDeletion) {
+        this.snapshotManager = snapshotManager;
+        this.tagManager = tagManager;
+        this.fileIO = fileIO;
+        this.snapshotDeletion = snapshotDeletion;
+        this.tagDeletion = tagDeletion;
+
+        this.tags = tagManager.tags();
+        this.tagManifestEntries = new HashMap<>();
+    }
+
+    /**
+     * Clean snapshots and tags whose id is larger than given snapshot's in the reverse direction.
+     */
+    public void cleanLargerThan(Snapshot snapshot) {
+        // snapshots are cleaned first because it will collect tag manifest entries
+        cleanSnapshotsLargerThan(snapshot);
+        cleanTagsLargerThan(snapshot);
+
+        // modify the latest hint
+        try {
+            snapshotManager.commitLatestHint(snapshot.id());
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private void cleanSnapshotsLargerThan(Snapshot snapshot) {
+        long snapshotId = snapshot.id();
+        List<Snapshot> toBeCleaned = new ArrayList<>();
+
+        long earliest =
+                checkNotNull(
+                        snapshotManager.earliestSnapshotId(), "Cannot find earliest snapshot.");
+        long latest =
+                checkNotNull(snapshotManager.latestSnapshotId(), "Cannot find latest snapshot.");
+
+        // delete snapshot files first, cannot be read now
+        // it is possible that some snapshots have been expired
+        long to = Math.max(earliest, snapshotId + 1);
+        for (long i = latest; i >= to; i--) {
+            toBeCleaned.add(snapshotManager.snapshot(i));
+            fileIO.deleteQuietly(snapshotManager.snapshotPath(i));
+        }
+
+        // delete data files
+        for (Snapshot s : toBeCleaned) {
+            cleanSnapshotDataFiles(s);
+        }
+
+        // delete directories
+        snapshotDeletion.cleanDataDirectories();
+
+        // delete manifest files
+        cleanManifests(snapshot, toBeCleaned, snapshotDeletion);
+    }
+
+    // don't concern about tag data files because file deletion methods won't throw exception
+    // when deleting non-existing data files
+    private void cleanSnapshotDataFiles(Snapshot snapshot) {
+        List<ManifestEntry> delta = new ArrayList<>();
+        snapshotDeletion
+                .tryReadManifestEntries(snapshot.deltaManifestList())
+                .iterator()
+                .forEachRemaining(delta::add);
+        if (tags.containsKey(snapshot)) {
```

Comment:
Maybe we can just let it be a good tag, we don't need to delete manifest and datafiles.